### PR TITLE
feature(font): draw img in label/span

### PR DESCRIPTION
**Pull request by guoweilkd**
### Description of the feature or fix

we can use lv_img_font  driver  to draw images in the  label/span obj

eg:
```c
lv_font_t * imgfont = lv_imgfont_create(80, get_imgfont_path);
if(imgfont == NULL) {
    LV_LOG_ERROR("imgfont init error");
}

imgfont->fallback = LV_FONT_DEFAULT;

lv_obj_t * label1 = lv_label_create(lv_scr_act());
lv_label_set_text(label1, "12\uF600\uF617AB");
lv_obj_set_style_text_font(label1, imgfont, LV_PART_MAIN);
lv_obj_center(label1);
```

image

### Checkpoints
- [x] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [x] Run `code-format.py` from the `scripts` folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [x] Update the documentation

### CODE REVIEW

File: docs/libs/img_font.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+```eval_rst
+.. include:: /header.rst 
+:github_url: |github_link_base|/libs/img_font.md
+```
+## img_font
```
Comment (kisvegabor) [inline docs/libs/img_font.md]:
```suggestion
## Image font (img_font)
```

File: docs/libs/img_font.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+Draw image in label or span obj with img_font.
+This is often used to display Unicode emoji icons in text.
+Supported image formats: determined by LVGL imge decoder.
```
Comment (kisvegabor) [inline docs/libs/img_font.md]:
```suggestion
Supported image formats: determined by LVGL image decoder.
```

File: docs/libs/img_font.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+## Usage
+Enable `LV_USE_IMG_FONT` in `lv_conf.h`.
+
```
Comment (kisvegabor) [inline docs/libs/img_font.md]: Some more info about how it works and the usage of `lv_img_font_t` would be great.

File: examples/libs/img_font/index.rst [hunk 4]
```diff
@@ -0,0 +1,6 @@
+Open a front with FreeType
```
Comment (kisvegabor) [inline examples/libs/img_font/index.rst]:
```suggestion
Use emojis in a text. 
```

File: src/extra/libs/img_font/lv_img_font.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+/**********************
+ *      TYPEDEFS
+ **********************/
+typedef struct lv_img_font_t {
+    /* gets the image path of the character */
+    bool (*get_img_path)(struct lv_img_font_t * handle, char * path,
+                         uint16_t len, uint32_t unicode, uint32_t unicode_next);
+    lv_font_t * font;   /* point to lvgl font */
+    uint16_t height;    /* font size */
+} lv_img_font_t;
```
Comment (kisvegabor) [inline src/extra/libs/img_font/lv_img_font.h]: Should have a 
```c
#if LV_USE_USER_DATA
void * user_date;
#endif
```

field for MicroPython compatibility.

File: examples/libs/img_font/lv_example_img_font_1.c [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    return true;
+}
+
+/**
+ * draw img in label or span obj
+ */
+void lv_example_img_font_1(void)
+{
+    static lv_img_font_t img_font;
+    img_font.font = NULL;
```
Comment (kisvegabor) [inline examples/libs/img_font/lv_example_img_font_1.c]: What about an `lv_img_font_init(height, get_img_cb)` function?

File: examples/libs/img_font/lv_example_img_font_1.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    img_font.height = 80;
+    img_font.get_img_path = get_img_path;
+
+    if(!lv_img_font_create(&img_font)) {
+        LV_LOG_ERROR("img font init error");
+    }
+
+#if LV_USE_FREETYPE
```
Comment (kisvegabor) [inline examples/libs/img_font/lv_example_img_font_1.c]: Can we copy e.g. `LV_FONT_DEFAULT` and the fallback there?
It's more general than using FreeType.

File: examples/libs/img_font/lv_example_img_font_1.c [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+ * draw img in label or span obj
+ */
+void lv_example_img_font_1(void)
+{
+    static lv_img_font_t img_font;
+    img_font.font = NULL;
+    img_font.height = 80;
+    img_font.get_img_path = get_img_path;
+
+    if(!lv_img_font_create(&img_font)) {
```
Comment (kisvegabor) [inline examples/libs/img_font/lv_example_img_font_1.c]: Returning a `lv_font_t *`?

File: src/draw/sw/lv_draw_sw_letter.c [hunk 9]
```diff
@@ -172,6 +172,22 @@ LV_ATTRIBUTE_FAST_MEM static void draw_letter_normal(lv_draw_ctx_t * draw_ctx, c
     uint32_t shades;
     if(bpp == 3) bpp = 4;
 
+    if(bpp == 7) { //is img
```
Comment (kisvegabor) [inline src/draw/sw/lv_draw_sw_letter.c]: I'd rather use `> 8` bpp  and define it for better readibility:
```c
#define LV_IMG_FONT_BPP  9
```

File: src/extra/libs/img_font/lv_img_font.c [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    LV_ASSERT_NULL(font);
+    img_dsc_t * dsc = (img_dsc_t *)font->dsc;
+    return (uint8_t *)dsc->path;
+}
+
+static bool img_font_get_glyph_dsc(const lv_font_t * font, lv_font_glyph_dsc_t * dsc_out,
+                                   uint32_t unicode_letter, uint32_t unicode_letter_next)
+{
+    img_dsc_t * dsc = (img_dsc_t *)font->dsc;
+    if(!dsc->info->get_img_path(dsc->info, dsc->path,
```
Comment (kisvegabor) [inline src/extra/libs/img_font/lv_img_font.c]:
```c
if(dsc->info->get_img_path == NULL) return false;
```

File: examples/libs/img_font/lv_example_img_font_1.c [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+LV_IMG_DECLARE(emoji_F617);
+static bool get_img_path(lv_img_font_t * handle, char * path, uint16_t len, uint32_t unicode, uint32_t unicode_next)
+{
+    LV_UNUSED(unicode_next);
+
+    if(unicode == 0xF617) {
+        memcpy((void *)path, &emoji_F617, sizeof(lv_img_dsc_t));
+    }
+    else {
+        snprintf(path, len, "%s/%04X.%s", "./lvgl/examples/assets/emoji", unicode, "png");
```
Comment (kisvegabor) [inline examples/libs/img_font/lv_example_img_font_1.c]: There should be a driver letter at the beginning of the path. E.g.
```
"A:lvgl/examples/assets/emoji"
```

File: src/extra/libs/img_font/lv_img_font.c [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+ *********************/
+#define LV_IMG_FONT_PATH_MAX_LEN 64
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+typedef struct {
+    lv_img_font_t * info;
+    char path[LV_IMG_FONT_PATH_MAX_LEN];
+} img_dsc_t;
```
Comment (kisvegabor) [inline src/extra/libs/img_font/lv_img_font.c]: Wouldn't be clearer to add an `lv_font_t` field here directly instead of in in `info`?

**Review by kisvegabor — COMMENTED** [review]
Thank you!

I was already thinking a lot about how to approach this question, but never thought of simple creating a font for it. Nice job! 

Had some comments but all of them are only minor things. 

What do you think about renaming this to `lv_imgfont`? Starting with `lv_img_` implies that it's related to the `lv_img` widget.

I think it should go the `src/extra/others` as it's not a 3rd party lib integration.

File: examples/libs/img_font/lv_example_img_font_1.c (hunk 6 above)
Comment (guoweilkd) [reply examples/libs/img_font/lv_example_img_font_1.c]: Is it better to use ` lv_font_t * lv_imgfont_create(height, get_img_cb) ` and keep it in pairs with `void lv_imgfont_destroy(lv_font_t * font)`

Comment (kisvegabor) [reply examples/libs/img_font/lv_example_img_font_1.c]: Sounds great!

File: examples/others/imgfont/lv_example_imgfont_1.c [hunk 13]
```diff
@@ -0,0 +1,52 @@
+#include "../../lv_examples.h"
+#include <stdio.h>
+
+#if LV_BUILD_EXAMPLES
+#if LV_USE_IMGFONT
+
+LV_IMG_DECLARE(emoji_F617);
+static bool get_imgfont_path(const lv_font_t * font, char * path,
```
Comment (kisvegabor) [inline examples/others/imgfont/lv_example_imgfont_1.c]:
```suggestion
static bool get_imgfont_path(const lv_font_t * font, void * img_src,
```

**Review by kisvegabor — COMMENTED** [review]
Looks good, thanks for the update!

Only one more minor thing...

**Comment by kisvegabor** [comment]
Looks good, thank you!

I just added imgfont to CI and fix some minor warnings.

**Pull request merged**